In [13]:
# Upload dataset.csv
from google.colab import files
uploaded = files.upload()

Saving house_price_prediction_dataset.csv to house_price_prediction_dataset (2).csv


In [14]:
#importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score


In [15]:
# Load the dataset
df = pd.read_csv("house_price_prediction_dataset.csv")
df.head(10)

,size_m2,year_built,city,bedrooms,bathrooms,distance_to_center_km,price_eur
0,132,1961,Stuttgart,1,1,13.66,925983
1,209,1940,Frankfurt,2,3,9.71,1527273
2,122,1956,Leipzig,5,3,8.42,507373
3,44,2024,Hamburg,2,2,13.23,320410
4,136,2020,Hamburg,3,3,11.67,929374
5,101,1959,Frankfurt,4,3,19.18,723543
6,218,1953,Dresden,2,2,24.49,711265
7,50,1969,Cologne,1,1,9.26,290224
8,132,2000,Leipzig,5,1,6.80,577855
9,151,1956,Stuttgart,3,1,20.17,885783


In [16]:
# Checking columns
print(df.columns)


Index(['size_m2', 'year_built', 'city', 'bedrooms', 'bathrooms',
       'distance_to_center_km', 'price_eur'],
      dtype='object')


In [17]:
# separating target(y) and features(x)
X = df.drop(columns=['price_eur'])
y = df['price_eur']

print("X shape:", X.shape)
print("y shape:", y.shape)

X.head(), y.head()


X shape: (5000, 6)
y shape: (5000,)


(   size_m2  year_built       city  bedrooms  bathrooms  distance_to_center_km
 0      132        1961  Stuttgart         1          1                  13.66
 1      209        1940  Frankfurt         2          3                   9.71
 2      122        1956    Leipzig         5          3                   8.42
 3       44        2024    Hamburg         2          2                  13.23
 4      136        2020    Hamburg         3          3                  11.67,
 0     925983
 1    1527273
 2     507373
 3     320410
 4     929374
 Name: price_eur, dtype: int64)

In [18]:
# Changing city to a numeric as linear regression cannot be applied on text values.
X = pd.get_dummies(X, columns=['city'], drop_first=True)
X.head()


,size_m2,year_built,bedrooms,bathrooms,distance_to_center_km,city_Cologne,city_Dresden,city_Dusseldorf,city_Frankfurt,city_Hamburg,city_Leipzig,city_Munich,city_Nuremberg,city_Stuttgart
0,132,1961,1,1,13.66,False,False,False,False,False,False,False,False,True
1,209,1940,2,3,9.71,False,False,False,True,False,False,False,False,False
2,122,1956,5,3,8.42,False,False,False,False,False,True,False,False,False
3,44,2024,2,2,13.23,False,False,False,False,True,False,False,False,False
4,136,2020,3,3,11.67,False,False,False,False,True,False,False,False,False


In [19]:
from sklearn.model_selection import train_test_split

# Spliting data: in 80%  for training, 20% for testing
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print("Training set size:", X_train.shape, y_train.shape)
print("Testing set size:", X_test.shape, y_test.shape)


Training set size: (4000, 14) (4000,)
Testing set size: (1000, 14) (1000,)


In [20]:
from sklearn.linear_model import LinearRegression

# Creating the linear regression model for the dataset
model = LinearRegression()

# Training the model using the training data
model.fit(X_train, y_train)

print("Model trained successfully!")


Model trained successfully!


In [21]:
# Predicting house prices using the trained model
y_pred = model.predict(X_test)

# Display first 10 predicted prices
y_pred[:10]


array([ 181300.43450568,  571715.39263458,  923577.4537508 ,
       1298290.48751379, 1691191.82057329,  767352.90166397,
        700624.33704376,  869167.50129343,  619360.73439408,
        659079.83907925])

In [22]:
from sklearn.metrics import mean_squared_error, r2_score

# Calculating error and accuracy of the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Mean Squared Error:", mse)
print("R² Score:", r2)


Mean Squared Error: 10452603673.261047
R² Score: 0.9420862354470145


In [23]:
# Creating a dummy house input (same feature format as training data)

dummy_house = pd.DataFrame([[
    120,   # size_m2
    2015,  # year_built
    3,     # bedrooms
    2,     # bathrooms
    6,     # distance_to_center_km
    0, 0, 0, 0, 0, 0, 1, 0, 0  # city columns (Munich = 1)
]], columns=X.columns)

# Predicting price for the dummy house
predicted_price = model.predict(dummy_house)

print("Predicted House Price (€):", round(predicted_price[0], 2))


Predicted House Price (€): 1147979.36


In [24]:
	# testing on the same dataset value to understand the variation.

dummy_house = pd.DataFrame([[
    132	,   # size_m2
    1961,  # year_built
    1,     # bedrooms
    1,     # bathrooms
    13.66,     # distance_to_center_km
    0, 0, 0, 0, 0, 0, 0, 0, 1  # city columns (Stuttgart = 1)
]], columns=X.columns)

# Predicting price for the dummy house
predicted_price = model.predict(dummy_house)

print("Predicted House Price (€):", round(predicted_price[0], 2))


Predicted House Price (€): 905281.0


In [25]:
	# testing on the same dataset value to understand the variation.

dummy_house = pd.DataFrame([[
    232	,   # size_m2
    1961,  # year_built
    1,     # bedrooms
    1,     # bathrooms
    13.66,     # distance_to_center_km
    0, 0, 0, 0, 0, 0, 0, 0, 1  # city columns (Stuttgart = 1)
]], columns=X.columns)

# Predicting price for the dummy house
predicted_price = model.predict(dummy_house)

print("Predicted House Price (€):", round(predicted_price[0], 2))

Predicted House Price (€): 1482137.17


1. Why did you choose the model you worked with, given this dataset?
- I chose the model: Linear Regression as my dataset that I create was mainly consisting of the numerical values  for example house size, year built, number of bedrooms, bathrooms, and distance from the city center.
- This model is simple to understand and easy to implement, which made it suitable for this applying on my dataset.
- House prices usually increase or decrease as per the size, year of built and location, so as per this i found linear regression  as the best fit.

2. Which feature do you believe influenced your predictions the most, and why?
- Size of the house (size_m2) influenced the predictions the most.As from real-life experience, larger houses usually cost more.
- Also, during prediction, changing the size value caused significant changes in the predicted price, which shows that the model heavily depends on this feature.

3. Do you personally trust the predictions produced by your model? Explain.
- Yes, I trust the predictions, but not completely.My model achieved an R² score of about 94%, which means it explains most of the price variation.
but, my model is trained only on limited features and rows, so the model predictions are good, but not the exact values, just close to the original prices.
- Thats why, I trust the predictions for approximate pricing, but not for final real-world decisions.

4. What important real-world information is missing from the dataset that could change the results?
- Important missing features that can affect the result are :-
    1. Condition of the house (new, renovated, old)
    2. Floor number and presence of an elevator
    3. Parking availability
    4. Nearby facilities like schools, hospitals, and public transport
    5. View, balcony, or garden
These are the few factors that i think can strongly affect house prices, and including them would likely improve model accuracy.

5. If your model performed poorly, what would be the first thing you would try to improve, and why?
- The first thing I would try is adding better and more useful features to the dataset.
- I understood that the model can only learn from the data we give it, so if important details are missing, the predictions will not be very accurate.
- will try to include more real-world information like house condition, location details, and nearby facilities, because these things clearly affect house prices.
- I would also try other regression models like Decision Trees or k-NN to compare results and see which performs better.
